api/openapi-spec/swagger.json

```diff
@@ -17553,13 +17553,26 @@
      }
     }
    },
+   "v1.InterfaceBindingMigration": {
+    "type": "object",
+    "properties": {
+     "method": {
```

Comment:
I find it odd that while we have `networkAttachmentDefinition` pointing directly to the namespace/name format and `sidecarImage` pointing to the image, we now have a nested object for `migration`.

Any reason to break the consistency? I would expect either:

```json
                {"passt": {
                    "networkAttachmentDefinition": "default/netbindingpasst",
                    "sidecarImage": "registry:5000/kubevirt/network-passt-binding:devel",
                    "migrationMethod": "link-refresh"
                    }
                }
```

or

```json
                {"passt": {
                    "networkAttachmentDefinition": {
                        "namespace": "default",
                        "name": "netbindingpasst"
                    },
                    "sidecar": {
                        "image": "registry:5000/kubevirt/network-passt-binding:devel"
                    },
                    "migration": {
                        "method": "link-refresh"
                    }
                }
```

Reply:
I don't mind changing it to the flat version. @EdDev @ormergi ?

Reply:
Q: What is the pro of:
```yaml
obj:
  migration: {}
```
A: It requires less fields and still can be extended (by adding `method` and other). It saves the need to add enabled field.
Its absence (`migration: nil`) is considered implicitly `disabled`.

Q: What are the pros to be explicit with:
```yaml
obj:
  migration:
    enabled: true
```
A: We can look at the API from two directions:
- Client: From this direction, the client wants to express that it "does not care" about this option, therefore not specifying it. If the client does mind, it will explicitly set it with the enable or disable directive.
- Server: From this direction, the absence if the fields/section implies the use of a default. This default may be "natural" (e.g. disabled), dependent on others (e.g. a default configuration value) or just less "natural" (e.g. enabled).

We should probably raise this also in sig-api , but I tend to prefer the explicit option.

I'd prefer the above points are explicitly answered to before deciding.
I think the current suggestion is wrong, too similar to the legacy binding options we have and are causing a mess.

Reply:
@phoracek Now I recall why the flat option wasn't good enough for migration.
We may have a migratable plugin (that kubevirt is not blocking migration when a VM has an interface using it) but this plugin may not have any special `migrationMethod`.

Reply:
@EdDev 
what would
```
obj:
  migration:
    method: "link-refresh" 
```
mean? enabled or not? I believe that the fact that we should handle defaults make the suggested approach (with enabled field) more complex and less intuitive.
From the client perspective - the client have to explicitly say whether the plugin is migratable or not. In what scenario the client doesn't care about it? 
From the server perspective- same. kubevirt doesn't know whether the plugin is migratable or not, any default my be wrong.

Reply:
I wonder if we should make 'method' more flexible.
For example, the SR-IOV binding require perperation berofe the migration start and when it finished (attach/detach host-devices and memolck adjustments).
Maybe its worth adding additional fields to set strategies for different migration phases, or set the 'method' field as slice to enable performing more then one operation.

WDYT?

Reply:
One method can reflect the before and after operation. The API got a green light before starting the PR, please lets proceed with it.
It is marked as alpha and behind a feature gate, if we find out the API is dramaticly wrong we can change it on the next release.

Reply:
I couldn't find example such as ours in the doc; a field that accept empty struct as value (e.g.: `pod: {}`).
It similar to a boolean field but instead of `true` the value has to be empty struct, and for `false` omitted the field from the spec.

The doc does mention bool fields https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types 
- last bullet.
Seems that this PR change follow the convention of bool fields.

Reply:
> Seems that this PR change follow the convention of bool fields.

How is an empty struct follows any convention there? The only thing I see mentioned is that instead of using a bool use a string with a meaningful value, like we do for the method field.

Reply:
> The book refer to k8s API convention doc

And the other book is using the API, with practical usages. I guess both are valid, but the one we actually should really follow (as we use it) is the builder.

Reply:
> How is an empty struct follows any convention there? The only thing I see mentioned is that instead of using a bool use a string with a meaningful value, like we do for the method field.

If you consider empty struct as bool, which seem to be a variation of bool field, then its like the mention bullet, in the sense we use string as value..

> And the other book is using the API, with practical usages. I guess both are valid, but the one we actually should 
really follow (as we use it) is the builder.

The book doesn't discuss API conventions much, it focus on the lifecycle of controllers and CRDs (creating a controller, defining CRD, extending them, versioning etc.., more like machinery stuff).

Reply:
> If you consider empty struct as bool, which seem to be a variation of bool field, then its like the mention bullet, in the sense we use string as value..

I am trying to point out that there is no such pattern or example in the books that consider an empty struct as a bool. And it is not an empty struct that is considered a book, but its absence (i.e. `nil` value) that is considered `false`. In the book a `nil` value is considered a default, with potential other parameters set on it.